services/slam/fake/slam.go

```diff
@@ -0,0 +1,160 @@
+	"go.viam.com/utils"
+	"go.viam.com/utils/artifact"
+
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/pointcloud"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/rimage"
+	"go.viam.com/rdk/services/slam"
+	"go.viam.com/rdk/spatialmath"
+	rdkutils "go.viam.com/rdk/utils"
+	"go.viam.com/rdk/vision"
+)
+
+var model = resource.NewDefaultModel("fake")
+
+const (
+	internalStateTemplate = "slam/example_cartographer_outputs/internal_state/internal_state_%d.pbstream"
+	maxDataCount          = 16
+	pcdTemplate           = "slam/example_cartographer_outputs/pointcloud/pointcloud_%d.pcd"
+	pngTemplate           = "slam/example_cartographer_outputs/image_map/image_map_%d.png"
+	positionTemplate      = "slam/example_cartographer_outputs/position/position_%d.txt"
+)
+
+func init() {
+	registry.RegisterService(
+		slam.Subtype,
+		model,
+		registry.Service{
+			Constructor: func(
+				ctx context.Context,
+				_ registry.Dependencies,
+				config config.Service,
+				logger golog.Logger,
+			) (interface{}, error) {
+				return &SLAM{Name: config.Name, dataCount: 1}, nil
+			},
+		},
+	)
+}
+
+var _ = slam.Service(&SLAM{})
+
+// SLAM is a fake slam that returns generic data.
+type SLAM struct {
+	generic.Echo
+	Name      string
+	dataCount int
+}
+
+func (slamSvc *SLAM) GetMap(ctx context.Context, name, mimeType string, cp *referenceframe.PoseInFrame,
+	include bool, extra map[string]interface{},
+) (string, image.Image, *vision.Object, error) {
+	var err error
+	var img image.Image
+	var vObj *vision.Object
+
+	// Increment data after getMap call
+	slamSvc.incrementDataCount()
+
+	switch mimeType {
+	case rdkutils.MimeTypePCD:
+		f, err := os.Open(artifact.MustPath(fmt.Sprintf(pcdTemplate, slamSvc.dataCount)))
+		if err != nil {
+			return "", nil, nil, err
+		}
+		defer utils.UncheckedErrorFunc(f.Close)
+		pc, err := pointcloud.ReadPCDToBasicOctree(f)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		vObj, err = vision.NewObject(pc)
+		if err != nil {
+			return "", nil, nil, err
+		}
+
+	case rdkutils.MimeTypeJPEG:
+		img, err = rimage.NewImageFromFile(artifact.MustPath(fmt.Sprintf(pngTemplate, slamSvc.dataCount)))
+
+	default:
+		return "", nil, nil, errors.New("received invalid mimeType for GetMap call")
+	}
+
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return mimeType, img, vObj, nil
+}
+
+func (slamSvc *SLAM) Position(ctx context.Context, name string, extra map[string]interface{}) (*referenceframe.PoseInFrame, error) {
+	data, err := os.ReadFile(artifact.MustPath(fmt.Sprintf(positionTemplate, slamSvc.dataCount)))
+	if err != nil {
+		return nil, err
+	}
+
+	substrings := strings.Split(string(data), " | ")
+
+	point, err := extract(strings.Split(substrings[0], " "))
+	if err != nil {
+		return nil, err
+	}
+	xyz := r3.Vector{X: point[0], Y: point[1], Z: point[2]}
+
+	orientations, err := extract(strings.Split(substrings[1], " "))
```

Comment:
Does the file contain axis-angle representation or orientation vector?

Reply:
This is how Jeremy generated the dataset: https://github.com/jeremyhyde-viam/rdk/commit/bc06b04b7a8d06b560c7fe8f7439a6a0d3040b12#diff-041927d429c5b9c2868ecfc5c12d0aa719ea98978dad15f65ab76f25c5440435R594

& uses `p.Orientation().AxisAngles().RX` `p.Orientation().AxisAngles().RY` `p.Orientation().AxisAngles().RZ` `p.Orientation().AxisAngles().Theta`

So I assume orientation vector?

Reply:
Also, sorry but I don't understand the distinction between the two.

Reply:
If this is calling `AxisAngles()` then it's using axis angles instead of the orientation vector coefficients. Slam GetPose is supposed to return a pose, which includes an orientation vector. See this doc for the difference: https://docs.google.com/document/d/1kPj6WqJmm1x5WJkI6LwWZE9fPQZDK3TFIXcB5YtTplg/edit. And documentation for pose: https://github.com/viamrobotics/api/blob/782f2d07a2c1b67709e703a49459ec53eba5af2d/proto/viam/common/v1/common.proto#L32

Reply:
I've renamed the variables to denote axis angles & called `spatialmath.NewR4AA().AxisAngles()` before setting the struct members, so I believe I'm now setting the pose using the same representation as the archive data, but please correct me if I've misunderstood.

Reply:
Yes, this looks correct to me, if the archive data is using axis angle representation.